src/libstd/io/lazy.rs

```diff
@@ -26,7 +26,11 @@ const fn done<T>() -> *mut Arc<T> { 1_usize as *mut _ }
 unsafe impl<T> Sync for Lazy<T> {}
 
 impl<T: Send + Sync + 'static> Lazy<T> {
-    pub const fn new(init: fn() -> Arc<T>) -> Lazy<T> {
+    /// Safety: `init` must not call `get` on the variable that is being
+    /// initialized.
+    pub const unsafe fn new(init: fn() -> Arc<T>) -> Lazy<T> {
+        // `lock` is never initialized fully, so it is UB to attempt to
```

Comment:
This threw me for a spin when I first read it, could it perhaps be reworded? In the Rust sense we fully initialize the lock (aka put valid bytes in place) but what this comment is alluding to is the POSIX-specific behavior tweak in the mutex implementation where `Mutex::new` isn't as safe as an `init`'d mutex. Could this perhaps be reworded as:

> `lock` is not valid to use in a reentrant fashion, causing UB if that happens;
> for more information see comments in <path>